### PR TITLE
Decouple replication lag from logic to fail stale replicas

### PR DESCRIPTION
**Pull request by ankitkala**
### Description
Current implementation relies on the replication timer tracked by primary's checkpoint tracker to evaluation replica's staleness. While this correctly provide the replication_lag, it is not ideal to rely on replication lag as it also includes the timer taken by primary shard to upload the data to remote store. We shouldn't penalize & fail replicas if the primary is slow in uploading the segments. To get around this, we need to track two times separately: 
- replication lag: total time taken by replica to catchup after primary refreshes.
- replication time:total time taken by replica to catchup after primary publishes the checkpoint(excludes the time to upload to remote store)

Changes done: 
- Added a wrapper timer object which tracks 2 separate times(when timer was created v/s when timer started)
- Modified the logic for `indexShard.updateReplicationCheckpoint` to only create the timers (and not start)
- Added hook to start the timers after checkpoint is published. 
- Used the replication time for failing stale replicas in SegRep backpressure service.
- Introduced a setting to configure the staleness time limit before primary starts failing stale replica shards.
- Updated the Replication stats tracker to use the newer replication lag metric.

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/8453

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 1]
```diff
@@ -1208,24 +1215,25 @@ public synchronized void updateVisibleCheckpointForShard(final String allocation
     }
 
     /**
-     * After a new checkpoint is published, start a timer for each replica to the checkpoint.
+     * After a new checkpoint is published, create a timer for each replica to the checkpoint.
      * @param checkpoint {@link ReplicationCheckpoint}
      */
-    public synchronized void setLatestReplicationCheckpoint(ReplicationCheckpoint checkpoint) {
+    public synchronized void setLatestReplicationCheckpoint(ReplicationCheckpoint checkpoint, boolean shouldStartTimers) {
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1221]: i don't think we need the boolean flag here - shard can invoke create/start separately?

Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java]: Yep, removed the flag.

File: server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    private long creationTime;
+
+    public SegmentReplicationLagTimer() {
+        super();
+        creationTime = System.nanoTime();
+    }
+
+    public SegmentReplicationLagTimer(StreamInput in) throws IOException {
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java:37]: Is this only to be consistent with ReplicationTimer as writeable?   I don't believe we do/want to send these over the wire.

**Review by mch2 — COMMENTED** [review]
Overall looks good - test seems to be a bit faky? testFailStaleReplicaTaskDisabled

File: server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java (hunk 2 above)
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java:37]: Yep, just for consistency. Don't have any usecase where we'd send it over wire

**Comment by ankitkala** [comment]
Flaky `mixedClusterTest`: https://github.com/opensearch-project/OpenSearch/issues/9547

**Comment by ankitkala** [comment]
> Overall looks good - test seems to be a bit faky? testFailStaleReplicaTaskDisabled

Test failed but don't see any issue in the logs. Also ran the test over 13K iterations. Haven't failed again. I'l follow up incase i see more occurances.

File: server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java [hunk 3]
```diff
@@ -65,13 +66,20 @@ public class SegmentReplicationPressureService implements Closeable {
         Setting.Property.NodeScope
     );
 
-    public static final Setting<TimeValue> MAX_REPLICATION_TIME_SETTING = Setting.positiveTimeSetting(
+    public static final Setting<TimeValue> MAX_REPLICATION_TIME_BACKPRESSURE_SETTING = Setting.positiveTimeSetting(
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java:69]: can you pls add a comment here on the difference between this and `MAX_REPLICATION_LIMIT_STALE_REPLICA_SETTING`

File: server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java [hunk 4]
```diff
@@ -65,13 +66,20 @@ public class SegmentReplicationPressureService implements Closeable {
...
         "segrep.pressure.time.limit",
         TimeValue.timeValueMinutes(5),
         Setting.Property.Dynamic,
         Setting.Property.NodeScope
     );
 
+    public static final Setting<TimeValue> MAX_REPLICATION_LIMIT_STALE_REPLICA_SETTING = Setting.positiveTimeSetting(
+        "segrep.replication.time.limit",
+        TimeValue.timeValueMinutes(15),
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java:78]: I'm wondering the usefulness of setting this to 15m over simply turning it off by default?

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java [hunk 5]
```diff
@@ -1236,8 +1236,7 @@ private void startReplicationLagTimers() {
                     && replicationGroup.getUnavailableInSyncShards().contains(allocationId) == false
                     && latestReplicationCheckpoint.isAheadOf(cps.visibleReplicationCheckpoint)) {
                     cps.checkpointTimers.computeIfAbsent(latestReplicationCheckpoint, ignored -> {
-                        final ReplicationTimer replicationTimer = new ReplicationTimer();
-                        replicationTimer.start();
+                        final SegmentReplicationLagTimer replicationTimer = new SegmentReplicationLagTimer();
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1239]: nit return new SegmentReplicationLagTimer()

File: server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import java.io.IOException;
+
+/**
+ * Wrapper class for Replication Timer which also tracks time elapsed since the timer was created.
+ * Currently, this is being used to calculate
+ * 1. Replication Lag: Total time taken by replica to sync after primary refreshed.
+ * 2. Replication event time: Total time taken by replica to sync after primary published the checkpoint
+ *                     (excludes the time spent by primary for uploading the segments to remote store).
+ */
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java:23]: pls add @opensearch.internal annotation

File: server/src/main/java/org/opensearch/index/SegmentReplicationShardStats.java [hunk 7]
```diff
@@ -30,6 +30,7 @@ public class SegmentReplicationShardStats implements Writeable, ToXContentFragme
     private final long checkpointsBehindCount;
     private final long bytesBehindCount;
     private final long currentReplicationTimeMillis;
```
Comment (mch2) [inline server/src/main/java/org/opensearch/index/SegmentReplicationShardStats.java:32]: can you pls add a comment on whats different between time & lag here? Or rename `currentReplicationTimeMillis` to `currentOngoingReplicationTimeMillis`

Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/SegmentReplicationShardStats.java:32]: Added the comment.

File: server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java (hunk 6 above)
Comment (ankitkala) [reply server/src/main/java/org/opensearch/indices/replication/common/SegmentReplicationLagTimer.java:23]: Thanks. Done.

File: server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java (hunk 5 above)
Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/seqno/ReplicationTracker.java:1239]: Ack.

File: server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java (hunk 3 above)
Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java:69]: Sure.

File: server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java (hunk 4 above)
Comment (ankitkala) [reply server/src/main/java/org/opensearch/index/SegmentReplicationPressureService.java:78]: I think we did discuss this offline. Let's keep it to off by default.

**Review by mch2 — APPROVED** [review]

**Pull request merged**